[w32handle] Fix duplicated handles for WaitAny and WaitAll

jaykrell opened the pull request:
Fixes https://github.com/mono/mono/issues/9089

kumpera commented on mono/tests/wait-dup-all.cs:
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Threading;
+
+class Program
```
Those tests would be better included in the corlib test suite. mono/tests is better for tests that must be standalone or require custom process setup.

jaykrell replied:
This area of the tree is much easier for me to work with, which I realize is not the best reason. Editing the makefile, editing-existing/adding-new/running tests, all much easier here.

kumpera commented on mono/metadata/threads.c:
```diff
@@ -2085,16 +2086,59 @@ ves_icall_System_Threading_WaitHandle_Wait_internal (gpointer *handles, gint32 n
+		// In particular, it is triggered by an error, and it is handled-based
+		// instead of object-based.
+		// See coreclr/src/vm/threads.cpp CheckForDuplicateHandles and its use.
+		if (win32WaitResult == WAIT_FAILED
```
I'm not 100% sure of whether we need to keep this weird split on error detection.
Can't we just have mono_win32_wait_for_multiple_objects_ex return a `MONO_W32HANDLE_WAIT_RET_FAILED_DUPLICATES`?

@luhenry what do you think?

jaykrell replied:
The existing win32 layer is very thin & so I left it alone after brief consideration, but yes that might be a better structure.

kumpera replied:
Tests in mono/tests takes a LOT longer to run compared to the unit tests in corlib.

Plus if you're making changes that are easily visible to C#, you should be running that test suite anyways.

jaykrell commented:
There is a bigger problem here on Win32 actually.

```
MonoOSEventWaitRet
mono_os_event_wait_multiple (MonoOSEvent **events, gsize nevents, gboolean waitall, guint32 timeout, gboolean alertable)
{
...
	else if (res == WAIT_FAILED)
		g_error ("%s: mono_thread_win32_wait_multiple_handle failed with error %d", __func__, 
}
```

We either need to revise this layer, or do the check earlier.

jaykrell commented on mono/metadata/w32handle.c:
```diff
@@ -939,29 +947,46 @@ mono_w32handle_wait_multiple (gpointer *handles, gsize nhandles, gboolean waital
+						if (handles_data [i] == handles_data [j]) {
+							mono_w32handle_unref (handles_data [j]);
+							handles_data [j] = NULL;
+						}
```
Actually that data isn't sorted, and operating on the sorted data is tricky, so the code is a bit inefficient but about what it should be. Duplicates will be rare. Worst case is 64x64 but usually much less. The initial check for duplicates will tend to make this entirely skipped, esp. since mono didn't handle this before.

luhenry replied:
You have 2 choices here to avoid the inner-loop:
 - if growing upward (from i = 1 to nhandles), you can just set the `handles_data_sorted [i - 1]` to `NULL` so if `handles_data_sorted [i + 1]` is also a duplicate you haven't changed `handles_data_sorted [i]`
 - if growing downward (from i = nhandles - 1 to 0), you can still set `handles_data_sorted [i]` to `NULL` when comparing it to `handles_data_sorted [i + 1]` so when looking at `handles_data_sorted [i - 1]`, `handles_data_sorted [i]` haven't been changed.

I would go for the first one since it's the minimal change and the result is the same.

luhenry replied:
I missed the fact that you are checking `handles_data_sorted` but modifying `handles_data`. Extracting the duplication removal would certainly help in making the code clearer. In simpler terms: please move https://github.com/jaykrell/mono/blob/4493213cd1cc722603e838eb1e5b9977d404ff91/mono/metadata/w32handle.c?utf8=✓#L964-L988 after https://github.com/mono/mono/pull/9114/files#diff-81316870ec3e337f2fd9f434dfbd0a4cR962 loop, using a `duplicated` variable for example.

jaykrell replied:
There is really no need to move it. We can do the "action" right after the "detection". There is only one instance of each so no need to flow multiple  "detections" into the one "action". If a variable is introduced, the human reader has to go searching for all the reads and writes, only to find one of each and wonder why the code is split apart. The comments explain everything pretty well here I think.

luhenry replied:
Well, I didn't. What makes sense to you might not makes sense to other people and vice-versa.

jaykrell commented:
Ok, I moved the duplicate checking down a later. I used optional MonoError* instead of a new disposition. Optionality feels much safer than letting the new code run in other paths, even if it is rare and an error path, it still has new non-required side effects.

I also moved the tests as requested but haven't managed to run them yet. Running stuff in that part of the tree requires building a bunch more stuff that I haven't succeeded i yet -- trying again from the root, which is known not to work well but might leave enough success behind it.

jaykrell commented:
[comment was in wrong PR]
Note these are two slightly different layers:
`mono_win32_wait_for_multiple_objects_ex`
`mono_w32handle_wait_multiple`

The second one does not exist on Windows, and contains the first.
Nevertheless, they are quite similar, and adding a MonoError* to both should work.

jaykrell commented:
Three failures. All interpreter. Merits more investigation.

jaykrell commented:
oh, wrong PR. maybe flakes. Maybe. Must look closer.

jaykrell commented:
Need to fix windows build.

jaykrell commented:
@monojenkins build failed

luhenry commented on mono/metadata/w32handle.c:
```diff
@@ -131,6 +130,8 @@ mono_w32handle_unlock_signal_mutex (void)
 void
 mono_w32handle_lock (MonoW32Handle *handle_data)
 {
+	if (!handle_data)
```
This cannot transparently fail. Check in the caller that we do not lock a invalid handle.

jaykrell replied:
Ignoring null is not failing. Let me see  if it works, there might have been another problem.

luhenry replied:
It fails to do the expected thing, thus breaking assumptions.

luhenry commented on mono/metadata/w32handle.c:
```diff
@@ -143,6 +144,8 @@ mono_w32handle_trylock (MonoW32Handle *handle_data)
 void
 mono_w32handle_unlock (MonoW32Handle *handle_data)
 {
+	if (!handle_data)
```
This cannot transparently fail. Check in the caller that we do not unlock a invalid handle.

luhenry commented on mono/metadata/w32handle.c:
```diff
@@ -418,6 +421,9 @@ mono_w32handle_ref_core (MonoW32Handle *handle_data)
 static gboolean
 mono_w32handle_unref_core (MonoW32Handle *handle_data)
 {
+	if (!handle_data)
```
This cannot transparently fail. Check in the caller that we do not unref a invalid handle.

luhenry commented on mono/utils/mono-error-internals.h:
```diff
@@ -241,6 +241,9 @@ mono_error_set_null_reference (MonoError *error)
 }
 
+// FIXME static inline failed to link for Windows
+#define mono_error_set_duplicate_wait_object(error) (mono_error_set_generic_error (error, "System", "DuplicateWaitObjectException", "Duplicate objects in argument."))
```
Have that function like the rest of the other functions: the declaration in this file and the implementation in `mono-error.c`

jaykrell replied:
I'll try that. The link error was surprising.

jaykrell replied:
Problem was missing the include. I not used to the C behavior of allowing to call unprototyped functions and error not until link time but I'll remember it now. At least on Windows -- maybe there is a warning we can/should elevate for that, like the other compilers.

jaykrell commented:
@monojenkins build failed

jaykrell commented:
Oh, the windows break was  probably lack of include.
Lack of prototype isn't enough to error.

jaykrell commented:
@monojenkins build failed

jaykrell commented:
Related-seeming windows failure

jaykrell commented:
Maybe previous structure of fix better but will look more.

jaykrell commented:
@monojenkins build failed

jaykrell commented:
That does not make sense to me. 

The loop is nested in order to find all duplicates of all handles, once we find there are any. Not just the duplicates of the first duplicate found.

Nulling the sorted data does not work, because wait must return the lowest signaled index, against the original unsorted handles. Nulling the sorted data does not help so easily. You’d have to map back to the unsorted data, which is possible but does not seem desirable.

jaykrell commented:
@monojenkins build failed

jaykrell commented:
ok now?

jaykrell commented:
CI passed. Approve?

jaykrell commented:
@monojenkins build failed

jaykrell commented:
Approve?

jaykrell commented:
Review?

jaykrell commented:
Review?

jaykrell commented:
Approve?

jaykrell commented:
Approve?

lateralusX commented on mono/metadata/w32handle.c:
```diff
@@ -895,14 +902,75 @@ mono_w32handle_wait_one (gpointer handle, guint32 timeout, gboolean alertable)
+
+	MonoW32Handle *sorted [MONO_W32HANDLE_MAXIMUM_WAIT_OBJECTS]; // 64
+	memcpy (sorted, handles, nhandles * sizeof (handles[0]));
+	qsort (sorted, nhandles, sizeof (gpointer), g_direct_equal);
```
Just for clarity and consistency, `qsort (sorted, nhandles, sizeof  (sorted[0]), g_direct_equal);`

lateralusX commented on mono/utils/mono-os-wait-win32.c:
```diff
@@ -121,29 +123,48 @@ mono_win32_wait_for_single_object_ex (HANDLE handle, DWORD timeout, BOOL alertab
+			&& GetLastError () == ERROR_INVALID_PARAMETER) {
+		gpointer handles_sorted [MAXIMUM_WAIT_OBJECTS]; // 64
+		memcpy (handles_sorted, handles, count * sizeof (handles [0]));
+		qsort (handles_sorted, count, sizeof (gpointer), g_direct_equal);
```
Just for clarity and consistency, `qsort (handles_sorted, count, sizeof  (handles_sorted [0]), g_direct_equal);`

jaykrell commented:
@monojenkins build Linux AArch64 Coop Suspend

sourceskyboxer commented:
And llvm looks failed since "}" error if you see if make compilation throws error. Please fix it!

jaykrell commented:
Please provide more details. I don't see the error.
